ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java

```diff
@@ -115,8 +127,89 @@ public void start() throws ReplicationException {
+    } catch (InterruptedException e) {
+      logger.warn("Waiting for startup is interrupted.");
+    }
+    if (started) {
+      // Read-write lock avoids contention between addReplica()/removeReplica() and onReplicaAddedOrRemoved() methods.
+      // Read lock for current method should suffice because multiple threads from cluster change handlers should be able
+      // to access partitionToPartitionInfo map. Each thead only updates PartitionInfo of certain partition and synchronization
+      // is only required within PartitionInfo. Also, addRemoteReplicaInfoToReplicaThread() is thread-safe which allows
+      // several threads from cluster change handlers to add remoteReplicaInfo
+      rwLock.readLock().lock();
+      try {
+        // 2. determine if added/removed replicas have peer replica on local node.
+        //    We skip the replica on current node because it should already be added/removed by state transition thread.
+        Set<ReplicaId> addedPeerReplicas = addedReplicas.stream()
+            .filter(r -> partitionToPartitionInfo.containsKey(r.getPartitionId()) && r.getDataNodeId() != currentNode)
+            .collect(Collectors.toSet());
+        Set<ReplicaId> removedPeerReplicas = removedReplicas.stream()
+            .filter(r -> partitionToPartitionInfo.containsKey(r.getPartitionId()) && r.getDataNodeId() != currentNode)
+            .collect(Collectors.toSet());
+
+        // No additional synchronization is required because cluster change handler of each dc only updates replica-threads
+        // belonging to certain dc. Hence, there is only one thread adding/removing remote replicas within a certain dc.
+
+        // 3. create replicaInfo for new remote replicas and assign them to replica-threads.
+        List<RemoteReplicaInfo> replicaInfosToAdd = new ArrayList<>();
+        for (ReplicaId remoteReplica : addedPeerReplicas) {
+          PartitionInfo partitionInfo = partitionToPartitionInfo.get(remoteReplica.getPartitionId());
+          // create findToken, remoteReplicaInfo
+          FindToken findToken =
+              this.tokenHelper.getFindTokenFactoryFromReplicaType(remoteReplica.getReplicaType()).getNewFindToken();
+          RemoteReplicaInfo remoteReplicaInfo =
+              new RemoteReplicaInfo(remoteReplica, partitionInfo.getLocalReplicaId(), partitionInfo.getStore(),
+                  findToken,
+                  TimeUnit.SECONDS.toMillis(storeConfig.storeDataFlushIntervalSeconds) * Replication_Delay_Multiplier,
+                  SystemTime.getInstance(), remoteReplica.getDataNodeId().getPortToConnectTo());
+          logger.info("Adding remote replica {} on {} to partition info.", remoteReplica.getReplicaPath(),
+              remoteReplica.getDataNodeId());
+          if (partitionInfo.addReplicaInfoIfAbsent(remoteReplicaInfo)) {
+            replicationMetrics.addMetricsForRemoteReplicaInfo(remoteReplicaInfo);
+          }
+          replicaInfosToAdd.add(remoteReplicaInfo);
```

Comment:
we should only add remoteReplicaInfo to the "toAdd" list when it was successfully added to partitionInfo, so we should move this statement in the if-else statement above.

Reply:
Correct, will fix this.